subsys/net/l2/ethernet/arp.c

```diff
@@ -466,6 +473,138 @@ static void arp_gratuitous(struct net_if *iface,
 	}
 }
 
+#if defined(CONFIG_NET_ARP_GRATUITOUS_TRANSMISSION)
+static void arp_gratuitous_send(struct net_if *iface,
+				struct in_addr *ipaddr)
+{
+	struct net_arp_hdr *hdr;
+	struct net_pkt *pkt;
+
+	pkt = net_pkt_alloc_with_buffer(iface, sizeof(struct net_arp_hdr),
+					AF_UNSPEC, 0, NET_BUF_TIMEOUT);
+	if (!pkt) {
+		return;
+	}
+
+	net_buf_add(pkt->buffer, sizeof(struct net_arp_hdr));
+	net_pkt_set_vlan_tag(pkt, net_eth_get_vlan_tag(iface));
+
+	hdr = NET_ARP_HDR(pkt);
+
+	hdr->hwtype = htons(NET_ARP_HTYPE_ETH);
+	hdr->protocol = htons(NET_ETH_PTYPE_IP);
+	hdr->hwlen = sizeof(struct net_eth_addr);
+	hdr->protolen = sizeof(struct in_addr);
+	hdr->opcode = htons(NET_ARP_REQUEST);
+
+	memcpy(&hdr->dst_hwaddr.addr, net_eth_broadcast_addr(),
+	       sizeof(struct net_eth_addr));
+	memcpy(&hdr->src_hwaddr.addr, net_if_get_link_addr(iface)->addr,
+	       sizeof(struct net_eth_addr));
+
+	net_ipv4_addr_copy_raw(hdr->dst_ipaddr, (uint8_t *)ipaddr);
+	net_ipv4_addr_copy_raw(hdr->src_ipaddr, (uint8_t *)ipaddr);
+
+	net_pkt_lladdr_src(pkt)->addr = net_if_get_link_addr(iface)->addr;
+	net_pkt_lladdr_src(pkt)->len = sizeof(struct net_eth_addr);
+
+	net_pkt_lladdr_dst(pkt)->addr = (uint8_t *)net_eth_broadcast_addr();
+	net_pkt_lladdr_dst(pkt)->len = sizeof(struct net_eth_addr);
+
+	NET_DBG("Sending gratuitous ARP pkt %p", pkt);
+
+	if (net_if_send_data(iface, pkt) == NET_DROP) {
+		net_pkt_unref(pkt);
+	}
+}
+
+static void notify_all_ipv4_addr(struct net_if *iface)
+{
+	struct net_if_ipv4 *ipv4 = iface->config.ip.ipv4;
+	int i;
+
+	if (!ipv4) {
+		return;
+	}
+
+	for (i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+		if (ipv4->unicast[i].is_used &&
+		    ipv4->unicast[i].address.family == AF_INET &&
+		    ipv4->unicast[i].addr_state == NET_ADDR_PREFERRED) {
+			arp_gratuitous_send(iface,
+					    &ipv4->unicast[i].address.in_addr);
+		}
+	}
+}
+
+static void iface_event_handler(struct net_mgmt_event_callback *cb,
+				uint32_t mgmt_event, struct net_if *iface)
+{
+	ARG_UNUSED(cb);
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		return;
+	}
+
+	if (mgmt_event != NET_EVENT_IF_UP) {
+		return;
+	}
+
+	notify_all_ipv4_addr(iface);
+}
+
+static void ipv4_event_handler(struct net_mgmt_event_callback *cb,
+			       uint32_t mgmt_event, struct net_if *iface)
+{
+	struct in_addr *ipaddr;
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
```

Comment:
We need to check also if the interface is VLAN one.
```
@@ -542,7 +542,8 @@ static void iface_event_handler(struct net_mgmt_event_callback *cb,
 {
        ARG_UNUSED(cb);
 
-       if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+       if (!(net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET) ||
+             net_eth_is_vlan_interface(iface))) {
                return;
        }
 
@@ -558,7 +559,8 @@ static void ipv4_event_handler(struct net_mgmt_event_callback *cb,
 {
        struct in_addr *ipaddr;
 
-       if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+       if (!(net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET) ||
+             net_eth_is_vlan_interface(iface))) {
                return;
        }
 
@@ -583,7 +585,8 @@ static void iface_cb(struct net_if *iface, void *user_data)
 {
        ARG_UNUSED(user_data);
 
-       if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+       if (!(net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET) ||
+             net_eth_is_vlan_interface(iface))) {
                return;
        }
```
This requires #71909 which contains the helper function returning this information, I had missed implementing it.

Reply:
Done